server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -0,0 +1,158 @@
+         * will be set to {@link org.elasticsearch.index.seqno.SequenceNumbers#UNASSIGNED_SEQ_NO} and the exception will be non-null. If the
+         * global checkpoint is updated, the exception will be null.
+         *
+         * @param globalCheckpoint the updated global checkpoint
+         * @param e                if non-null, the shard is closed
+         */
+        void accept(long globalCheckpoint, IndexShardClosedException e);
+    }
+
+    // guarded by this
+    private boolean closed;
+    private volatile List<GlobalCheckpointListener> listeners;
+    private volatile long lastKnownGlobalCheckpoint = UNASSIGNED_SEQ_NO;
+
+    private final ShardId shardId;
+    private final Executor executor;
+    private final Logger logger;
+
+    /**
+     * Construct a global checkpoint listeners collection.
+     *
+     * @param shardId  the shard ID on which global checkpoint updates can be listened to
+     * @param executor the executor for listener notifications
+     * @param logger   a shard-level logger
+     */
+    GlobalCheckpointListeners(
+            final ShardId shardId,
+            final Executor executor,
+            final Logger logger) {
+        this.shardId = Objects.requireNonNull(shardId);
+        this.executor = Objects.requireNonNull(executor);
+        this.logger = Objects.requireNonNull(logger);
+    }
+
+    /**
+     * Add a global checkpoint listener. If the global checkpoint is above the current global checkpoint known to the listener then the
+     * listener will fire immediately on the calling thread.
+     *
+     * @param currentGlobalCheckpoint the current global checkpoint known to the listener
+     * @param listener                the listener
+     */
+    synchronized void add(final long currentGlobalCheckpoint, final GlobalCheckpointListener listener) {
+        if (closed) {
+            throw new IllegalStateException("can not listen for global checkpoint changes on a closed shard [" + shardId + "]");
+        }
+        if (lastKnownGlobalCheckpoint > currentGlobalCheckpoint) {
+            // notify directly
+            executor.execute(() -> notifyListener(listener, lastKnownGlobalCheckpoint, null));
+        } else {
+            if (listeners == null) {
+                listeners = new ArrayList<>();
+            }
+            listeners.add(listener);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        synchronized (this) {
+            closed = true;
+        }
+        notifyListeners(UNASSIGNED_SEQ_NO, new IndexShardClosedException(shardId));
+    }
+
+    /**
+     * Invoke to notify all registered listeners of an updated global checkpoint.
+     *
+     * @param globalCheckpoint the updated global checkpoint
+     */
+    void globalCheckpointUpdated(final long globalCheckpoint) {
+        assert globalCheckpoint >= NO_OPS_PERFORMED;
+        lastKnownGlobalCheckpoint = globalCheckpoint;
```

Comment:
I think this has to be synchronized to avoid race conditions with add. Say an add method already sampled the `lastKnownGlobalCheckpoint` and concluded that it has to add a listener. It then goes on to update the `listeners` array but haven't done so yet (it's still null). The `notifyListeners` is called here and it runs the `listeners != null` test before the add method made it non null. It that case I think we miss the listener.

I may be missing something here, but my point is that this is too complex IMO. I don't see why we can't just make this a simple fully synchronized method. The listeners are call on executor anyway.